src/Periphery/ReceiverChainflip.sol

```diff
@@ -0,0 +1,154 @@
+        (
+            bytes32 transactionId,
+            LibSwap.SwapData[] memory swapData,
+            address receiver
+        ) = abi.decode(message, (bytes32, LibSwap.SwapData[], address));
+
+        // execute swap(s)
+        _swapAndCompleteBridgeTokens(
+            transactionId,
+            swapData,
+            token,
+            payable(receiver),
+            amount
+        );
+    }
+
+    /// Private Methods ///
+
+    /// @notice Performs a swap before completing a cross-chain transaction
+    /// @param _transactionId the transaction id associated with the operation
+    /// @param _swapData array of data needed for swaps
+    /// @param assetId address of the token received from the source chain
+    /// @param receiver address that will receive tokens in the end
+    /// @param amount amount of token
+    /// @notice Performs a swap before completing a cross-chain transaction
+    /// @param _transactionId The transaction id associated with the operation
+    /// @param _swapData Array of data needed for swaps
+    /// @param assetId Address of the token received from the source chain
+    /// @param receiver Address that will receive tokens in the end
+    /// @param amount Amount of token to swap
+    /// @dev If the swap fails, the original bridged tokens are sent directly to the receiver
+    /// @dev Approvals are reset to 0 after the operation completes
+    function _swapAndCompleteBridgeTokens(
+        bytes32 _transactionId,
+        LibSwap.SwapData[] memory _swapData,
+        address assetId,
+        address payable receiver,
+        uint256 amount
+    ) private {
+        // Convert Chainflip's native token address to LibAsset.NATIVE_ASSETID
+        address actualAssetId = assetId == CHAINFLIP_NATIVE_ADDRESS
+            ? LibAsset.NATIVE_ASSETID
+            : assetId;
+
+        // Don't need approval for native token
+        if (actualAssetId != LibAsset.NATIVE_ASSETID) {
+            actualAssetId.safeApproveWithRetry(address(executor), amount);
+        }
+
+        try
+            executor.swapAndCompleteBridgeTokens{
+                value: actualAssetId == LibAsset.NATIVE_ASSETID ? amount : 0
+            }(_transactionId, _swapData, actualAssetId, receiver)
+        {} catch {
+            // send the bridged (and unswapped) funds to receiver address
+            if (actualAssetId == LibAsset.NATIVE_ASSETID) {
+                // Handle native token using safeTransferETH
+                receiver.safeTransferETH(amount);
+            } else {
+                // Handle ERC20 token
+                actualAssetId.safeTransfer(receiver, amount);
+            }
+
+            emit LiFiTransferRecovered(
+                _transactionId,
+                actualAssetId,
+                receiver,
+                amount,
+                block.timestamp
+            );
+        }
+
+        // Only reset approval for non-native tokens
+        if (actualAssetId != LibAsset.NATIVE_ASSETID) {
```

Comment:
you can also include that....

We are checkling like 4 times or so if it's native asset or not....this can all be added to one if block